File: frontend/src/api/platformApi.js
import api from "./axiosConfig";

const PLATFORM_TOKEN_KEY = "gns-platform-token";
const PLATFORM_ADMIN_KEY = "gns-platform-admin";

export function getPlatformToken() {
  return localStorage.getItem(PLATFORM_TOKEN_KEY);
}

export function setPlatformSession({ access_token, admin }) {
  localStorage.setItem(PLATFORM_TOKEN_KEY, access_token);
  localStorage.setItem(PLATFORM_ADMIN_KEY, JSON.stringify(admin));
}

export function clearPlatformSession() {
  localStorage.removeItem(PLATFORM_TOKEN_KEY);
  localStorage.removeItem(PLATFORM_ADMIN_KEY);
}

export function getPlatformAdmin() {
  try {
    const raw = localStorage.getItem(PLATFORM_ADMIN_KEY);
    return raw ? JSON.parse(raw) : null;
  } catch {
    return null;
  }
}

function platformHeaders() {
  const token = getPlatformToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function withApiErrorHandling(label, operation, context = {}) {
  try {
    return await operation();
  } catch (error) {
    console.error(`[platformApi.${label}] Error performing request:`, error.message, context);
    throw error;
  }
}

export async function superAdminLogin(email, password) {
  return withApiErrorHandling(
    "superAdminLogin",
    async () => {
      const { data } = await api.post("/platform/auth/login", { email, password });
      return data;
    },
    { email }
  );
}

export async function superAdminVerifyOtp(challengeToken, otp) {
  return withApiErrorHandling(
    "superAdminVerifyOtp",
    async () => {
      const { data } = await api.post("/platform/auth/verify-otp", {
        challenge_token: challengeToken,
        otp,
      });
      return data;
    },
    { challengeToken }
  );
}

export async function superAdminResendOtp(challengeToken) {
  return withApiErrorHandling(
    "superAdminResendOtp",
    async () => {
      const { data } = await api.post("/platform/auth/resend-otp", {
        challenge_token: challengeToken,
      });
      return data;
    },
    { challengeToken }
  );
}

export async function getSuperAdminProfile() {
  return withApiErrorHandling(
    "getSuperAdminProfile",
    async () => {
      const { data } = await api.get("/platform/auth/me", { headers: platformHeaders() });
      return data;
    }
  );
}

export async function listCompanies() {
  return withApiErrorHandling(
    "listCompanies",
    async () => {
      const { data } = await api.get("/platform/companies", { headers: platformHeaders() });
      return data;
    }
  );
}

export async function createCompany(payload) {
  return withApiErrorHandling(
    "createCompany",
    async () => {
      const { data } = await api.post("/platform/companies", payload, { headers: platformHeaders() });
      return data;
    },
    { payload }
  );
}

export async function getCompany(tenantId) {
  return withApiErrorHandling(
    "getCompany",
    async () => {
      const { data } = await api.get(`/platform/companies/${tenantId}`, { headers: platformHeaders() });
      return data;
    },
    { tenantId }
  );
}

export async function updateCompany(tenantId, payload) {
  return withApiErrorHandling(
    "updateCompany",
    async () => {
      const { data } = await api.put(`/platform/companies/${tenantId}`, payload, {
        headers: platformHeaders(),
      });
      return data;
    },
    { tenantId, payload }
  );
}

export async function activateCompany(tenantId) {
  return withApiErrorHandling(
    "activateCompany",
    async () => {
      const { data } = await api.post(`/platform/companies/${tenantId}/activate`, null, {
        headers: platformHeaders(),
      });
      return data;
    },
    { tenantId }
  );
}

export async function suspendCompany(tenantId) {
  return withApiErrorHandling(
    "suspendCompany",
    async () => {
      const { data } = await api.post(`/platform/companies/${tenantId}/suspend`, null, {
        headers: platformHeaders(),
      });
      return data;
    },
    { tenantId }
  );
}

export async function deleteCompany(tenantId) {
  return withApiErrorHandling(
    "deleteCompany",
    async () => {
      await api.delete(`/platform/companies/${tenantId}`, { headers: platformHeaders() });
    },
    { tenantId }
  );
}

export async function resetCompanyPassword(tenantId, newPassword) {
  return withApiErrorHandling(
    "resetCompanyPassword",
    async () => {
      const { data } = await api.post(
        `/platform/companies/${tenantId}/reset-password`,
        { new_password: newPassword },
        { headers: platformHeaders() }
      );
      return data;
    },
    { tenantId }
  );
}

export async function listCompanyUsers(tenantId) {
  return withApiErrorHandling(
    "listCompanyUsers",
    async () => {
      const { data } = await api.get(`/platform/companies/${tenantId}/users`, {
        headers: platformHeaders(),
      });
      return data;
    },
    { tenantId }
  );
}

export async function getCompanySubscription(tenantId) {
  return withApiErrorHandling(
    "getCompanySubscription",
    async () => {
      const { data } = await api.get(`/platform/companies/${tenantId}/subscription`, {
        headers: platformHeaders(),
      });
      return data;
    },
    { tenantId }
  );
}

export async function updateCompanyLicense(tenantId, payload) {
  return withApiErrorHandling(
    "updateCompanyLicense",
    async () => {
      const { data } = await api.put(`/platform/companies/${tenantId}/license`, payload, {
        headers: platformHeaders(),
      });
      return data;
    },
    { tenantId, payload }
  );
}

File: backend/app/api/platform_api.py
"""GNS Super Admin platform API — OTP auth + company management."""

from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.api.platform_deps import get_current_super_admin
from app.models.platform import PlatformSuperAdmin
from app.schemas.platform import (
    CreateCompanyRequest,
    CreateCompanyResponse,
    ResetCompanyPasswordRequest,
    SuperAdminAuthResponse,
    SuperAdminLoginChallengeResponse,
    SuperAdminLoginRequest,
    SuperAdminResendOtpRequest,
    SuperAdminVerifyOtpRequest,
    UpdateCompanyRequest,
    UpdateLicenseRequest,
)
from app.middleware.security import check_rate_limit
from app.services.platform_company_service import PlatformCompanyService
from app.services.super_admin_service import SuperAdminService
from app.services.address_lookup_service import lookup_indian_pincode
from app.utils.api_response import success_response

router = APIRouter(prefix="/platform", tags=["platform"])


def _client_ip(request: Request) -> str | None:
    forwarded = request.headers.get("X-Forwarded-For")
    if forwarded:
        return forwarded.split(",")[0].strip()
    return request.client.host if request.client else None


@router.get("/address/pincode/{pincode}")
def platform_lookup_pincode(
    pincode: str,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
):
    """Same address lookup service for Super Admin company provisioning."""
    data = lookup_indian_pincode(pincode)
    return success_response("Address details retrieved", data)

@router.post("/auth/login", response_model=SuperAdminLoginChallengeResponse)
def super_admin_login(
    req: SuperAdminLoginRequest,
    request: Request,
    db: Session = Depends(get_db),
):
    """Validate Super Admin credentials, then send OTP to registered mobile."""
    try:
        return SuperAdminService(db).initiate_login(
            req.email,
            req.password,
            ip_address=_client_ip(request),
            user_agent=request.headers.get("User-Agent"),
        )
    except HTTPException:
        raise
    except SQLAlchemyError as exc:
        db.rollback()
        raise HTTPException(status_code=503, detail="Database service unavailable") from exc
    except Exception as exc:
        db.rollback()
        raise HTTPException(status_code=500, detail="Failed to initiate login") from exc


@router.post("/auth/verify-otp", response_model=SuperAdminAuthResponse)
def super_admin_verify_otp(
    req: SuperAdminVerifyOtpRequest,
    request: Request,
    db: Session = Depends(get_db),
):
    """Verify 6-digit OTP and issue platform JWT (user_id, company_id, role)."""
    try:
        return SuperAdminService(db).verify_login_otp(
            req.challenge_token,
            req.otp,
            ip_address=_client_ip(request),
            user_agent=request.headers.get("User-Agent"),
        )
    except HTTPException:
        raise
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except SQLAlchemyError as exc:
        db.rollback()
        raise HTTPException(status_code=503, detail="Database service unavailable") from exc
    except Exception as exc:
        db.rollback()
        raise HTTPException(status_code=500, detail="Failed to verify OTP") from exc


@router.post("/auth/resend-otp", response_model=SuperAdminLoginChallengeResponse)
def super_admin_resend_otp(
    req: SuperAdminResendOtpRequest,
    request: Request,
    db: Session = Depends(get_db),
):
    """Invalidate previous OTP and send a new one (60s cooldown + rate limit)."""
    try:
        return SuperAdminService(db).resend_otp(
            req.challenge_token,
            ip_address=_client_ip(request),
            user_agent=request.headers.get("User-Agent"),
        )
    except HTTPException:
        raise
    except SQLAlchemyError as exc:
        db.rollback()
        raise HTTPException(status_code=503, detail="Database service unavailable") from exc
    except Exception as exc:
        db.rollback()
        raise HTTPException(status_code=500, detail="Failed to resend OTP") from exc


@router.get("/auth/me")
def super_admin_me(
    admin: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return SuperAdminService(db).get_profile(admin.id)


@router.get("/companies")
def list_companies(
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).list_companies()


@router.post("/companies", response_model=CreateCompanyResponse, status_code=201)
def create_company(
    payload: CreateCompanyRequest,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).create_company(payload)


@router.get("/companies/{tenant_id}")
def get_company(
    tenant_id: int,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).get_company(tenant_id)


@router.put("/companies/{tenant_id}")
def update_company(
    tenant_id: int,
    payload: UpdateCompanyRequest,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).update_company(tenant_id, payload)


@router.post("/companies/{tenant_id}/activate")
def activate_company(
    tenant_id: int,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).activate_company(tenant_id)


@router.post("/companies/{tenant_id}/suspend")
def suspend_company(
    tenant_id: int,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).suspend_company(tenant_id)


@router.delete("/companies/{tenant_id}", status_code=204)
def delete_company(
    tenant_id: int,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    PlatformCompanyService(db).delete_company(tenant_id)


@router.post("/companies/{tenant_id}/reset-password")
def reset_company_password(
    tenant_id: int,
    payload: ResetCompanyPasswordRequest,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).reset_company_admin_password(
        tenant_id, payload.new_password
    )


@router.get("/companies/{tenant_id}/users")
def list_company_users(
    tenant_id: int,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).list_company_users(tenant_id)


@router.get("/companies/{tenant_id}/subscription")
def get_company_subscription(
    tenant_id: int,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).get_subscription(tenant_id)


@router.put("/companies/{tenant_id}/license")
def update_company_license(
    tenant_id: int,
    payload: UpdateLicenseRequest,
    _: PlatformSuperAdmin = Depends(get_current_super_admin),
    db: Session = Depends(get_db),
):
    return PlatformCompanyService(db).update_license(tenant_id, payload)
